libbeat/outputs/shipper/shipper.go

```diff
@@ -109,11 +111,28 @@ func (c *shipper) Connect() error {
 	if err != nil {
 		return fmt.Errorf("shipper connection failed with: %w", err)
 	}
-	c.log.Debugf("connect to %s established.", c.config.Server)
 
 	c.conn = conn
 	c.client = sc.NewProducerClient(conn)
 
+	indexClient, err := c.client.PersistedIndex(ctx, &messages.PersistedIndexRequest{
+		PollingInterval: durationpb.New(0), // no need to subscribe, we need it only once
+	})
+	if err != nil {
+		return fmt.Errorf("failed to connect to the server: %w", err)
+	}
+	serverInfo, err := indexClient.Recv()
+	if err != nil {
+		return fmt.Errorf("failed to fetch server information: %w", err)
+	}
+	c.serverID = serverInfo.GetUuid()
```

Comment:
The first call to `PublishEvents` actually requires the ID as one of the request parameters and the server validates it against its own value.

Reply:
I thought the ID was optional? The intention was that equality was only enforced if the client included the ID, so that clients connecting for the first time (or clients that don't care about ACKs) could omit it

Reply:
It is optional: https://github.com/elastic/elastic-agent-shipper/blob/34571c41e892f8498d817a3b2784cc372b0985a2/server/server.go#L108

We can omit it in the first call as Fae described, we don't care if the shipper restarts between this point and the first call to `PublishEvents` (at least not from the perspective of acknowledgement tracking).

Reply:
I think this can be removed to simplify the connection logic. The first `PublishEvents` RPC can have an empty server ID and everything will work correctly. We need to save the shipper ID return from that first call though.

Reply:
I'm not sure about this. `Connect` can be called multiple times by the beats pipeline:

https://github.com/elastic/beats/blob/91775be8b2592ec6995ee339b17e39fda65d7f5d/libbeat/publisher/pipeline/client_worker.go#L133-L159

This means that every time we have a connection error, we would re-use a previous serverID and the next batch publication would fail with the validation error which would cause an infinite loop if I removed this logic from `Connect`. I think it's more robust this way.

Reply:
Ah I didn't realize that just reading the code here in isolation, can you add a comment clarifying this?

Reply:
I'm not sure I follow the conflict here -- can't `Connect` just set `c.serverID` to an empty string and skip the whole `PersistedIndex` call? That will mean if there's a connection error, the server id will be reset, which is all we really need to keep the id from being reused, and `Publish` calls need to check the returned id anyway.

Reply:
@rdner I think this is the only unresolved thread, was it addressed with the most recent changes?

I approved as I don't think this is blocking given that we have some follow up changes coming.